packages/discount-code-generator/src/cli.js

```diff
@@ -0,0 +1,238 @@
+  )
+  .showHelpOnFail(false, 'Use --help to display the CLI options.\n')
+
+  .option('help', {
+    alias: 'h',
+  })
+  .help('help', 'Show help text.')
+
+  .option('version', {
+    alias: 'v',
+    type: 'boolean',
+  })
+  .version('version', 'Show version number.', version)
+
+  .option('quantity', {
+    alias: 'q',
+    describe: 'Quantity of discount codes to generate. (Between 1 and 500000)',
+    demandOption: true,
+  })
+  .coerce('quantity', (arg) => {
+    const quantity = parseInt(arg, 10)
+    // Limit quantity to 500000 to avoid `out-of-memory` error
+    if (quantity <= 0 || quantity > 500000)
+      throw new Error('Invalid quantity, must be a number between 1 and 500000')
+
+    return quantity
+  })
+
+  .option('code-length', {
+    alias: 'l',
+    default: 11,
+    describe: 'Length of the discount codes to generate.',
+  })
+  .coerce('code-length', arg => parseInt(arg, 10))
+
+  .option('code-prefix', {
+    alias: 'p',
+    default: '',
+    describe: 'Prefix for each code. No prefix will be used if omitted.',
+  })
+
+  .option('input', {
+    alias: 'i',
+    describe: 'Path to code options CSV or JSON file.',
+  })
+  .coerce('input', (arg) => {
+    if (fs.existsSync(arg)) {
+      if (arg.match(/\.json$/i) || arg.match(/\.csv$/i))
+        return fs.createReadStream(String(arg))
+
+      throw new Error('Invalid input file format. Must be CSV or JSON')
+    }
+    throw new Error('Input file cannot be reached or does not exist')
+  })
+
+  .option('output', {
+    alias: 'o',
+    default: 'stdout',
+    describe: 'Path to store generated output file.',
+  })
+  .coerce('output', (arg) => {
+    if (arg === 'stdout') {
+      npmlog.stream = fs.createWriteStream('discountCodeGenerator.log')
+      return process.stdout
+    }
+    if (arg.match(/\.json$/i) || arg.match(/\.csv$/i))
+      return fs.createWriteStream(String(arg))
+
+    throw new Error('Invalid output file format. Must be CSV or JSON')
+  })
+
+  .option('delimiter', {
+    alias: 'd',
+    default: ',',
+    describe: 'Used CSV delimiter for input and/or output file.',
+  })
+
+  .option('multivalueDelimiter', {
+    alias: 'm',
+    default: ';',
+    describe: 'Used CSV delimiter in multivalue fields for input/output file.',
+  })
+
+  .option('logLevel', {
+    default: 'info',
+    describe: 'Logging level: error, warn, info or verbose.',
+  })
+  .coerce('logLevel', (arg) => {
+    npmlog.level = arg
+  })
+  .argv
+
+// Resolve stream input to javascript object
+const resolveInput = (_args) => {
+  const input = _args.input
+  let _attributes
+  return new Promise((resolve, reject) => {
+    if (input === undefined)
+      resolve({})
+    else if (input.path.match(/\.json$/i))
+      // Convert input stream to JSON file
+      input
+        .on('error', (error) => {
+          reject(error)
+        })
+        .on('data', (data) => {
+          _attributes = data
```

Comment:
Shouldn't this be `_attributes += data` because the input file is loaded in chunks?